### ISSUE TITLE
[Sema] fails with type declared on following line

### ISSUE THREAD

**Issue by clementval** [flang-compiler/f18#929]
The following program does not pass the debug semantic pass. The following error is reported: 

```bash
./f18 dummy.f90
./dummy.f90:14:23: error: No explicit type declared for 'n'
    subroutine sub1(x,y,n,a)
                        ^
./install/bin/f18: semantic errors in ./dummy.f90
```

```fortran
program main
  implicit none
  integer, parameter :: n = 1024
  real :: x(n), y(n), a = 10.0

  x(:) = 1.0
  y(:) = 2.0

  call sub1(x,y,n,a)
  print*,sum(y(:))

contains

  subroutine sub1(x,y,n,a)
    implicit none
    real :: a, x(n), y(n)
    integer :: n, i
    do i = 1, n
      y(i) = a*x(i)+y(i)
    enddo
  end subroutine sub1

end program
```

If the `integer :: n, I` is placed before the `real :: a, x(n), y(n)` declaration the semantic checks pass without problem.

**Comment by tskeith** [comment]
I believe this program is not conforming. `n` is used in a specification expression in `sub1`. In 10.1.11 paragraph 6 it says:
> A variable in a specification expression shall have its type and type parameters, if any, specified by a previous declaration in the same scoping unit, by the implicit typing rules in effect for the scoping unit, or by host or use association. If a variable in a specification expression is typed by the implicit typing rules, its appearance in any subsequent type declaration statement shall confirm the implied type and type parameters.

**Comment by tskeith** [comment]
Please reopen if you disagree with this analysis.

**Comment by clementval** [comment]
@tskeith I agree with your analysis. Other compilers let this kind of code go through so that's why I opened the issue. What is the basic rule regarding other compilers difference to the Fortran standard?
